File: typescript/packages/core/src/store/approvals.ts
import { z } from "zod";
import { Int, type KnownEvent, principalKey } from "../log";
import type { Strict } from "../log/zod-types";
import { err, ok, type Result } from "../result";
import type { ChainEvent } from "../verify";
import { type LogError, logError } from "../verify/error";
import type { Tx } from "./driver";
import type { BranchRow } from "./tables";
import { parseRows } from "./tables";

// The approvals table (store.sql): one single-use row per approval_requested,
// written in that append's transaction, and consumed by the append of its answer. The log's own
// rule 9 makes a challenge single-use along one chain; the row makes it single-use across every
// branch that inherited it, and binds it to the branch it was asked on.

const STATES = ["open", "granted", "denied", "expired"] as const;

const ApprovalRow: Strict<{
  branch_id: z.ZodString;
  call_id: z.ZodString;
  args_hash: z.ZodString;
  expires_at: typeof Int;
  state: z.ZodEnum<{ [K in (typeof STATES)[number]]: K }>;
}> = z.strictObject({
  branch_id: z.string(),
  call_id: z.string(),
  args_hash: z.string(),
  expires_at: Int,
  state: z.enum(STATES),
});
export type ApprovalRow = z.infer<typeof ApprovalRow>;

export async function approvalRow(
  tx: Tx,
  challengeId: string,
): Promise<Result<ApprovalRow | undefined, LogError>> {
  const rows = parseRows(
    ApprovalRow,
    await tx.all(
      "SELECT branch_id, call_id, args_hash, expires_at, state FROM approvals WHERE challenge_id = ?",
      [challengeId],
    ),
  );
  return rows.ok ? ok(rows.value[0]) : rows;
}

type Answer = Extract<
  KnownEvent,
  { type: "approval_granted" | "approval_denied" }
>;

/**
 * Runs inside an append's transaction, after its events are inserted: opens a row for each
 * approval_requested and consumes the row of each answer. A row that is not open, or bound to
 * another branch or call, rolls the append back. A challenge with no row (one imported with
 * its log) is checked by the log's rules alone.
 */
export async function recordApprovals(
  tx: Tx,
  branch: BranchRow,
  added: readonly ChainEvent[],
  now: number,
): Promise<Result<void, LogError>> {
  for (const line of added) {
    if (line.kind !== "event") continue;
    const e = line.event;
    if (e.type === "approval_requested") await open(tx, branch, e.data);
    if (e.type !== "approval_granted" && e.type !== "approval_denied") continue;
    const consumed = await consume(tx, branch, e, now);
    if (!consumed.ok) return consumed;
  }
  return ok(undefined);
}

/**
 * The approval rows a verified log implies, from its settlement events alone, never the clock
 * (import): a challenge with a later answer is that answer's state, one without is open, and an
 * expired one is still refused when someone decides it.
 */
export async function projectApprovals(
  tx: Tx,
  branch: Owner,
  events: readonly KnownEvent[],
): Promise<void> {
  for (const e of events) {
    if (e.type === "approval_requested") await open(tx, branch, e.data);
    if (e.type === "approval_granted" || e.type === "approval_denied")
      await decide(tx, e, e.time);
  }
}

type Owner = Pick<BranchRow, "tenant_id" | "thread_id" | "branch_id">;

async function open(
  tx: Tx,
  branch: Owner,
  data: Extract<KnownEvent, { type: "approval_requested" }>["data"],
): Promise<void> {
  await tx.run(
    `INSERT INTO approvals (challenge_id, tenant_id, installation_id, thread_id, branch_id,
      call_id, args_hash, file_hashes, expires_at, state, decided_by, decided_at)
      VALUES (?, ?, NULL, ?, ?, ?, ?, ?, ?, 'open', NULL, NULL)`,
    [
      data.challenge_id,
      branch.tenant_id,
      branch.thread_id,
      branch.branch_id,
      data.call_id,
      data.args_hash,
      new TextEncoder().encode("[]"),
      data.expires_at,
    ],
  );
}

async function consume(
  tx: Tx,
  branch: BranchRow,
  e: Answer,
  now: number,
): Promise<Result<void, LogError>> {
  const id = e.data.challenge_id;
  const row = await approvalRow(tx, id);
  if (!row.ok) return row;
  if (row.value === undefined) return ok(undefined);
  const refused = refusal(row.value, branch, e, now);
  if (refused !== undefined) return err(refused);
  await decide(tx, e, now);
  return ok(undefined);
}

async function decide(tx: Tx, e: Answer, at: number): Promise<void> {
  await tx.run(
    "UPDATE approvals SET state = ?, decided_by = ?, decided_at = ? WHERE challenge_id = ? AND state = 'open'",
    [
      e.type === "approval_granted" ? "granted" : "denied",
      principalKey(e.actor.principal),
      at,
      e.data.challenge_id,
    ],
  );
}

function refusal(
  row: ApprovalRow,
  branch: BranchRow,
  e: Answer,
  now: number,
): LogError | undefined {
  const id = e.data.challenge_id;
  if (row.state !== "open")
    return logError(
      row.state === "expired" ? "approval_expired" : "approval_duplicate",
      `challenge ${id} is ${row.state}`,
    );
  if (
    row.branch_id !== branch.branch_id ||
    row.call_id !== e.data.call_id ||
    row.args_hash !== e.data.args_hash
  )
    return logError(
      "approval_mismatch",
      `${e.type} does not match challenge ${id}`,
    );
  return row.expires_at <= now
    ? logError("approval_expired", `challenge ${id} expired`)
    : undefined;
}

File: python/src/threads/store/approvals.py
"""Single-use approval challenges (store.sql `approvals`).

A row is inserted in the transaction that appends its approval_requested, and consumed by one
conditional update in the transaction that appends approval_granted or approval_denied. The
row carries the bindings the log doesn't: the tenant and the channel installation.
"""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal

from threads.log import ApprovalRequestedEvent, BranchId, CallId, ParseError, ThreadId
from threads.store.companion import Companion
from threads.store.conn import Conn
from threads.store.sql import int_of, text_of
from threads.store.verify import StoredEvent

type Decided = Literal["granted", "denied"]


@dataclass(frozen=True, slots=True)
class Challenge:
    challenge_id: str
    installation_id: str | None
    thread_id: ThreadId
    branch_id: BranchId
    call_id: CallId
    args_hash: str
    expires_at: int
    state: str


def record(conn: Conn, tenant_id: str, events: Sequence[StoredEvent]) -> None:
    """One open row per approval_requested of an append, bound to the thread's channel
    installation when the thread is a channel conversation's (null otherwise)."""
    for event in events:
        if not isinstance(event, ApprovalRequestedEvent):
            continue
        data = event.data
        conn.execute(
            "INSERT INTO approvals (challenge_id, tenant_id, installation_id, thread_id, branch_id,"
            " call_id, args_hash, file_hashes, expires_at, state)"
            " VALUES (?, ?, (SELECT installation_id FROM channel_threads"
            " WHERE tenant_id = ? AND thread_id = ? LIMIT 1), ?, ?, ?, ?, ?, ?, 'open')",
            # ponytail: no referenced-file hashes yet ("[]"); bind them when calls declare files.
            (
                data.challenge_id,
                tenant_id,
                tenant_id,
                event.thread_id,
                event.thread_id,
                event.branch_id,
                data.call_id,
                data.args_hash,
                b"[]",
                data.expires_at,
            ),
        )


def find(conn: Conn, tenant_id: str, challenge_id: str) -> Challenge | None:
    """The tenant's challenge; another tenant's is not found."""
    row = conn.execute(
        "SELECT installation_id, thread_id, branch_id, call_id, args_hash, expires_at, state"
        " FROM approvals WHERE challenge_id = ? AND tenant_id = ?",
        (challenge_id, tenant_id),
    ).fetchone()
    if row is None:
        return None
    installation, thread, branch, call, args, expires, state = row
    return Challenge(
        challenge_id,
        None if installation is None else text_of(installation),
        ThreadId(text_of(thread)),
        BranchId(text_of(branch)),
        CallId(text_of(call)),
        text_of(args),
        int_of(expires),
        text_of(state),
    )


def expire(conn: Conn, challenge_id: str, now: int) -> None:
    """An answer at or after the expiry moves an open row to expired: a denial."""
    conn.execute(
        "UPDATE approvals SET state = 'expired', decided_at = ?"
        " WHERE challenge_id = ? AND state = 'open' AND expires_at <= ?",
        (now, challenge_id, now),
    )


def consume(challenge_id: str, state: Decided, principal_key: str, now: int) -> Companion:
    """The conditional update that makes the answer single-use: only an open, unexpired row
    moves, so a second answer (or a race) is approval_duplicate and appends nothing."""

    def update(conn: Conn, _events: Sequence[StoredEvent]) -> ParseError | None:
        done = conn.execute(
            "UPDATE approvals SET state = ?, decided_by = ?, decided_at = ?"
            " WHERE challenge_id = ? AND state = 'open' AND expires_at > ?",
            (state, principal_key, now, challenge_id, now),
        )
        if done.rowcount == 1:
            return None
        return ParseError("approval_duplicate", f"challenge {challenge_id} is already answered")

    return update
